README.md
# An arbitrary-precision Decimal class type for PHP

[![Latest Version on Packagist](https://img.shields.io/packagist/v/piggly/php-decimal.svg?style=flat-square)](https://packagist.org/packages/piggly/php-decimal) [![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE) 

This library is similar to [decimal.js](https://github.com/MikeMcl/decimal.js/), all business logic credits and copyrights go to [Michael Mclaughlin](https://github.com/MikeMcl) and yours [contributors](https://github.com/MikeMcl/decimal.js/graphs/contributors).

The `piggly/php-decimal` is fully compatible with PHP and so fast as **decimal.js** library. See below all features attached to it.

## Features

This library is mainly composed by `Decimal` class. A `Decimal` is composed by coefficients, exponent and sign. It can handle integer and float values with an arbitrary-precision. 

Here precision is specified in terms of significant digits rather than decimal places, and all calculations are rounded to the precision (similar to Python's decimal module) rather than just those involving division.

This library also adds the trigonometric functions, among others, and supports non-integer powers, which makes it a significantly larger library.

* Integers and floats;
* Simple but full-featured API;
* Replicates many of the native math methods;
* Also handles hexadecimal, binary and octal values;
* Faster and easier to use than another PHP libraries;
* No dependencies or requirements;
* Comprehensive documentation and test set.

## Usage

`Decimal` can handle integers, floats, strings and `Decimal` objects:

```php
$w = new Decimal(123);
$x = new Decimal(123.4567);
$y = new Decimal('123456.7e-3');
$z = new Decimal($x);
```

A value can also be in binary, hexadecimal or octal if the appropriate prefix is included:

```php
$w = new Decimal('0xff.f');		// '255.9375'
$x = new Decimal('0b10101100');	// '172'
```

A `Decimal` is immutable in the sense that it is not changed by its methods, always returning a new `object`.

```php
$x = new Decimal(0.3);

$x->minus(0.1);						// $x is still 0.3
$y = $x->minus(0.1)->minus(0.1);	// $x is still 0.3 and $y is 0.1
```

All methods which returns Decimal values can be chained.

```php
$x->dividedBy($y)->plus($z)->times(9)->floor();
$x->times('1.23456780123456789e+9')->plus(9876.5432321)->dividedBy('4444562598.111772')->ceil();
```

Many method names have a shorter alias.

```php
$x->squareRoot()->dividedBy($y)->toPower(3)->equals($x->sqrt()->div($y)->pow(3))	// true
$x->cmp($y->mod($z)->neg()) == 1 && x->comparedTo($y->modulo($z)->negated()) == 1	// true
```

There are many ways to convert a `Decimal` to an string,

```php
$x = new Decimal(255.5);

$x->toExponential(5)  // '2.55500e+2'
$x->toFixed(5)        // '255.50000'
$x->toPrecision(5)    // '255.50'
$x->valueOf()         // '255.5'
```

and almost all methods are available as `static`.

```php
Decimal::sqrtOf('6.98372465832e+9823')      // '8.3568682281821340204e+4911'
Decimal::powOf(2, 0.0979843)                // '1.0702770511687781839'
```

`Decimal` can handle with `INF` and `NAN` values.

```php
$x = new Decimal(INF);     // INF
$y = new Decimal(NAN);     // NAN
```

There are many methods to do any checks to `Decimal`.

```php
$x->isCountless()    // If it is INF or NAN
$x->isFinite()       // If it is a finite number
$x->isInfinite()     // If it is an infinite number
$x->isInt()          // If it is an integer
$x->isNaN()          // If it is NAN
$x->isNegative()     // If it is negative
$x->isNulled()       // If it is INF, NAN or Zero
$x->isPositive()     // If it is positive
$x->isZero()         // If it is zero
```

By the way, there is a `toFraction` method with an optional maximum denominator argument.

```php
$z = new Decimal(355);
$pi = $z->dividedBy(113);        // '3->1415929204'
$pi->toFraction();               // [ '7853982301', '2500000000' ]
$pi->toFraction(1000);           // [ '355', '113' ]
```

All calculations are rounded according to the number of significant digits and rounding mode specified by the `precision` and `rounding` properties of the `DecimalConfig` object.

Each `Decimal` class is associated to a `DecimalConfig`. It may be the `global` configuration, or a custom configuration to that specific decimal number.

For advanced usage, multiple `Decimal` can be created, each with their own independent configuration which applies to all `Decimal` numbers created from it.

```php
// Set the precision and rounding of the global instance, 
// applies to all Decimal objects without configurations attached to it.
DecimalConfig::instance()->set([ 'precision' => 5, 'rounding' => 4 ]);

$decimal9 = DecimalConfig::clone()->set([ 'precision' => 9, 'rounding' => 1 ]);

$x = new Decimal(5);
$y = new Decimal(5, $decimal9);

$x->div(3);     // '1.6667'
$y->div(3);     // '1.66666666'

// $decimal9 applies to all `Decimal` numbers 
// created from $y in this case
$y->div(3)->times(1.5) // '2.50000000'
```

The value of a `Decimal` object is stored in a floating point format in terms of its `digits`, `exponent` and `sign`.

```php
$x = new Decimal(-12345.67);

$x->getDigits();          // [ 12345, 6700000 ]    digits (base 10000000)
$x->getExponent();        // 4                     exponent (base 10)
$x->getSign();            // -1                    sign
```

For further information see the [API](docs/api.md) reference in the docs directory, for now you may go to [decimal.js API](https://mikemcl.github.io/decimal.js/) since this library is fully compatible with it.

## Installation

### Composer

1. At you console, in your project folder, type `composer require piggly/php-decimal`;
2. Don't forget to add Composer's autoload file at your code base `require_once('vendor/autoload.php);`.

### Manual install

1. Download or clone with repository with `git clone https://github.com/piggly-dev/php-decimal.git`;
2. After, goes to `cd /path/to/piggly/php-decimal`;
3. Install all Composer's dependencies with `composer install`;
4. Add project's autoload file at your code base `require_once('/path/to/piggly/php-decimal/vendor/autoload.php);`.

## Dependencies 

The library has the following external dependencies:

* PHP 7.3+.

## TODO

In code, there annotations `@todo` with some improvements that this library may need.

## Changelog

See the [CHANGELOG](CHANGELOG.md) file for information about all code changes.

## Testing the code

This library uses the PHPUnit. We carry out tests of all the main classes of this application.

```bash
vendor/bin/phpunit
```

> You must always run tests with PHP 7.3 and greater. Any changes at this library need to pass of all oldest and newests tests.

> **!!** Some tests are heavy, be careful while testing them, they may require huge memory available.

## Contributions

See the [CONTRIBUTING](CONTRIBUTING.md) file for information before submitting your contribution.

## Credits

- [Caique Araujo](https://github.com/caiquearaujo);
- [All contributors](../../contributors);

### [decimal.js](https://github.com/MikeMcl/decimal.js/)

- [Michael Mclaughlin at decimal.js](https://github.com/MikeMcl);
- [Contributors at decimal.js](https://github.com/MikeMcl/decimal.js/graphs/contributors).

## Support the project

Piggly Studio is an agency located in Rio de Janeiro, Brazil. If you like this library and want to support this job, be free to donate any value to BTC wallet `3DNssbspq7dURaVQH6yBoYwW3PhsNs8dnK` ❤.

## License

MIT License (MIT). See [LICENSE](LICENSE).

CHANGELOG.md
# Changelog

## `v1.0.0` - `2021-05-24`

* First release.

tests/DecimalTest.php
<?php
namespace Piggly\Test\Decimal;

use PHPUnit\Framework\TestCase;
use Piggly\Decimal\Decimal;
use Piggly\Decimal\DecimalConfig;

class DecimalTest extends TestCase
{
	protected $_config;

	protected function setUp () : void
	{
		$this->_config = (new DecimalConfig())->set([
			'precision' => 40,
			'rounding' => 4,
			'toExpNeg' => -7,
			'toExpPos' => 20,
			'maxE' => 9e15,
			'minE' => -9e15,
			'crypto' => false,
			'modulo' => 1
		]);
	}

	/**
	 * Assert if constructor is same as expected data.
	 *
	 * @test Expecting positive assertion
    * @dataProvider constructorProvider
	 * @param array $coefficient
	 * @param integer $exponent
	 * @param integer $sign
	 * @param Decimal|integer|float|string $n
	 * @return void
	 */
	public function testIsMatchingConstructor ( 
		array $coefficient, 
		int $exponent, 
		int $sign, 
		$n 
	)
	{
		$d = new Decimal($n, $this->_config);

		$this->assertSame(
			[
				'coefficient' => $coefficient,
				'exponent' => $exponent,
				'sign' => $sign,
			],
			[
				'coefficient' => $d->_d(),
				'exponent' => $d->_e(),
				'sign' => $d->_s(),
			]
		);
	}

	/**
	 * Assert if is matching the expected data.
	 *
	 * @test Expecting positive assertion
    * @dataProvider stringMatching
	 * @param array $coefficient
	 * @param integer $exponent
	 * @param integer $sign
	 * @param Decimal|integer|float|string $n
	 * @return void
	 */
	public function testIsMatchingWithExpected (
		string $expected,
		$n 
	)
	{ $this->assertEquals($expected, (new Decimal($n, $this->_config))->valueOf()); }

	/**
	 * Provider for testIsMatchingConstructor().
	 * @return array
	 */
	public function constructorProvider() : array
	{
		return [
			[[0], 0, 1, 0],
			[[0], 0, 1, -0], // TODO :: Need to solve it, SHOULD BE -1 SIGN
			[[1], 0, -1, -1],
			[[10], 1, -1, -10],

			[[1], 0, 1, 1],
			[[10], 1, 1, 10],
			[[100], 2, 1, 100],
			[[1000], 3, 1, 1000],
			[[10000], 4, 1, 10000],
			[[100000], 5, 1, 100000],
			[[1000000], 6, 1, 1000000],

			[[1], 7, 1, 10000000],
			[[10], 8, 1, 100000000],
			[[100], 9, 1, 1000000000],
			[[1000], 10, 1, 10000000000],
			[[10000], 11, 1, 100000000000],
			[[100000], 12, 1, 1000000000000],
			[[1000000], 13, 1, 10000000000000],

			[[1], 14, -1, -100000000000000],
			[[10], 15, -1, -1000000000000000],
			[[100], 16, -1, -10000000000000000],
			[[1000], 17, -1, -100000000000000000],
			[[10000], 18, -1, -1000000000000000000],
			[[100000], 19, -1, -10000000000000000000],
			[[1000000], 20, -1, -100000000000000000000],

			[[1000000], -1, 1, 1e-1],
			[[100000], -2, -1, -1e-2],
			[[10000], -3, 1, 1e-3],
			[[1000], -4, -1, -1e-4],
			[[100], -5, 1, 1e-5],
			[[10], -6, -1, -1e-6],
			[[1], -7, 1, 1e-7],

			[[1000000], -8, 1, 1e-8],
			[[100000], -9, -1, -1e-9],
			[[10000], -10, 1, 1e-10],
			[[1000], -11, -1, -1e-11],
			[[100], -12, 1, 1e-12],
			[[10], -13, -1, -1e-13],
			[[1], -14, 1, 1e-14],

			[[1000000], -15, 1, 1e-15],
			[[100000], -16, -1, -1e-16],
			[[10000], -17, 1, 1e-17],
			[[1000], -18, -1, -1e-18],
			[[100], -19, 1, 1e-19],
			[[10], -20, -1, -1e-20],
			[[1], -21, 1, 1e-21],

			[[9], 0, 1, '9'],
			[[99], 1, -1, '-99'],
			[[999], 2, 1, '999'],
			[[9999], 3, -1, '-9999'],
			[[99999], 4, 1, '99999'],
			[[999999], 5, -1, '-999999'],
			[[9999999], 6, 1, '9999999'],

			[[9, 9999999], 7, -1, '-99999999'],
			[[99, 9999999], 8, 1, '999999999'],
			[[999, 9999999], 9, -1, '-9999999999'],
			[[9999, 9999999], 10, 1, '99999999999'],
			[[99999, 9999999], 11, -1, '-999999999999'],
			[[999999, 9999999], 12, 1, '9999999999999'],
			[[9999999, 9999999], 13, -1, '-99999999999999'],

			[[9, 9999999, 9999999], 14, 1, '999999999999999'],
			[[99, 9999999, 9999999], 15, -1, '-9999999999999999'],
			[[999, 9999999, 9999999], 16, 1, '99999999999999999'],
			[[9999, 9999999, 9999999], 17, -1, '-999999999999999999'],
			[[99999, 9999999, 9999999], 18, 1, '9999999999999999999'],
			[[999999, 9999999, 9999999], 19, -1, '-99999999999999999999'],
			[[9999999, 9999999, 9999999], 20, 1, '999999999999999999999']
		];
	}

	
	/**
	 * Provider for testIsMatchingWithExpected().
	 * @return array
	 */
	public function stringMatching() : array
	{
		return [
			// Binary.
			['0', '0b0'],
			['0', '0B0'],
			['-5', '-0b101'],
			['5', '+0b101'],
			// ['1.5', '0b1.1'],
			// ['-1.5', '-0b1.1'],

			// ['18181', '0b100011100000101.00'],
			// ['-12.5', '-0b1100.10'],
			// ['343872.5', '0b1010011111101000000.10'],
			// ['-328.28125', '-0b101001000.010010'],
			// ['-341919.144535064697265625', '-0b1010011011110011111.0010010100000000010'],
			// ['97.10482025146484375', '0b1100001.000110101101010110000'],
			// ['-120914.40625', '-0b11101100001010010.01101'],
			// ['8080777260861123367657', '0b1101101100000111101001111111010001111010111011001010100101001001011101001'],

			// // Octal.
			// ['8', '0o10'],
			// ['-8.5', '-0O010.4'],
			// ['8.5', '+0O010.4'],
			// ['-262144.000000059604644775390625', '-0o1000000.00000001'],
			// ['572315667420.390625', '0o10250053005734.31'],

			// // Hex.
			// ['1', '0x00001'],
			// ['255', '0xff'],
			// ['-15.5', '-0Xf.8'],
			// ['15.5', '+0Xf.8'],
			// ['-16777216.00000000023283064365386962890625', '-0x1000000.00000001'],
			// ['325927753012307620476767402981591827744994693483231017778102969592507', '0xc16de7aa5bf90c3755ef4dea45e982b351b6e00cd25a82dcfe0646abb']
		];
	}

	/**
	 * Get random integer.
	 *
	 * @return int
	 */
	private function randInt () : int
	{ return (int)\floor( $this->random() * 0x20000000000000 / \pow(10, $this->random() * 16 | 0) ); }

	/**
	 * Get random between 0 and 1.
	 *
	 * @return float
	 */
	private function random () : float
	{ return \mt_rand() / \mt_getrandmax(); }

	/**
	 * Prepare random to base.
	 *
	 * @param integer $base
	 * @return array
	 */
	private function prepareRandom ( $base = 10 ) : array
	{
		$r = randInt();
		return [\strval($r), \strval(\base_convert($r, 10, $base))];
	}
}

src/Math.php
<?php
namespace Piggly\Decimal;

use RuntimeException;

/**
 * Handle native PHP Math functions to
 * control data which will be returned.
 * 
 * There "alias" prevent to use all 
 * transformations required in main code
 * at Decimal class.
 * 
 * It also important to ajust global 
 * functions when requiring different
 * permissions.
 *
 * @since 1.0.0
 * @package Piggly\Decimal
 * @subpackage Piggly\Decimal
 * @author Caique Araujo <caique@piggly.com.br>
 */
class Math
{
	/**
	 * Any number type to a float value.
	 *
	 * @param Decimal|double|float|integer $num
	 * @since 1.0.0
	 * @return int
	 */
	public static function asFloat ( $num ) : float
	{ 
		if ( $num instanceof Decimal )
		{ return $num->toFloat(); }

		return \floatval($num); 
	}

	/**
	 * Any number type to an integer value.
	 *
	 * @param Decimal|double|float|integer $num
	 * @since 1.0.0
	 * @return int
	 */
	public static function asInt ( $num ) : int
	{ 
		if ( $num instanceof Decimal )
		{ return $num->toInt(); }

		return \intval($num); 
	}

	/**
	 * Any number type to a string value.
	 *
	 * @param Decimal|double|float|integer $num
	 * @since 1.0.0
	 * @return string
	 */
	public static function asStr ( $num ) : string
	{ 
		if ( $num instanceof Decimal )
		{ return $num->toString(); }

		return \strval($num); 
	}

	/**
	 * Check 5 rounding digits if `repeating` 
	 * is null, 4 otherwise.
	 * 
	 * `repeating == null` if caller is `log` or `pow`,
	 * `repeating != null` if caller is `naturalLogarithm` or `naturalExponential`.
	 *
	 * @param array $d Digits.
	 * @param integer $i Length.
	 * @param integer $rm Rounding mode.
	 * @param bool $repeating
	 * @since 1.0.0
	 * @return bool
	 */
	public static function checkRoundingDigits (
		$d,
		$i,
		$rm,
		$repeating = null
	) : bool
	{
		//  Get the length of the first word of the array d.
		for ( $k = $d[0]; $k >= 10; $k /= 10 )
		{ --$i; }

		// Is the rounding digit in the first word of d?
		if ( --$i < 0 )
		{
			$i += Decimal::LOG_BASE;
			$di = 0;
		}
		else
		{
			$di = static::intCeil(($i+1)/Decimal::LOG_BASE);
			$i %= Decimal::LOG_BASE;
		}

		// i is the index (0 - 6) of the rounding digit.
		// E.g. if within the word 3487563 the first rounding digit is 5,
		// then i = 4, k = 1000, rd = 3487563 % 1000 = 563
		$k = static::intPow(10, Decimal::LOG_BASE - $i);
		$rd = isset($d[$di]) && $d[$di] !== 0 ? $d[$di] % $k : 0;

		if ( \is_null($repeating) )
		{
			if ( $i < 3 )
			{
				if ( $i == 0 )
				{ $rd = ($rd/100) | 0; }
				else if ( $i == 1 )
				{ $rd = ($rd/10) | 0; }

				$r = (
					($rm < 4 && $rd == 99999) || 
					($rm > 3 && $rd == 49999) ||
					$rd == 50000 || 
					$rd == 0
				);
			}
			else 
			{
				$digit = isset($d[$di + 1]) ? ($d[$di + 1] / $k / 100) : 0;

				$r = (
					(($rm < 4 && $rd + 1 == $k) || ($rm > 3 && $rd + 1 == $k / 2)) &&
					(($digit) == static::intPow(10, $i - 2) - 1)
				) ||
				(($rd == $k / 2 || $rd == 0) && (($digit) == 0));
			}
		}
		else
		{
			if ( $i < 4 )
			{
				if ( $i == 0 )
				{ $rd = ($rd/1000) | 0; }
				else if ( $i == 1 )
				{ $rd = ($rd/100) | 0; }
				else if ( $i == 2 )
				{ $rd = ($rd/10) | 0; }

				$r = (($repeating || $rm < 4) && $rd == 9999) 
						|| (!$repeating && $rm > 3 && $rd == 4999);
			}
			else 
			{
				$digit = isset($d[$di + 1]) ? ($d[$di + 1] / $k / 1000) : 0;

				$r = ((($repeating || $rm < 4) && $rd + 1 == $k) 
						|| (!$repeating && $rm > 3 && $rd + 1 == $k / 2)) 
						&& ($digit) == static::intPow(10, $i - 3) - 1;
			}
		}
		
		return $r;
	}

	/**
	 * Convert string of $baseIn to an array of 
	 * numbers of $baseOut.
	 *
	 * Eg. convertBase('255', 10, 16) returns [15, 15].
	 * Eg. convertBase('ff', 16, 10) returns [2, 5, 5].
	 * 
	 * @param string $str
	 * @param integer $baseIn
	 * @param integer $baseOut
	 * @since 1.0.0
	 * @return array
	 */
	public static function convertBase (
		string $str,
		int $baseIn,
		int $baseOut
	) : array
	{
		$arr = [0];
		$i = 0;
		$strL = \strlen($str);

		for (; $i < $strL;)
		{
			for ( $arrL = \count($arr); $arrL--; )
			{ $arr[$arrL] *= $baseIn; }

			$arr[0] += \intval(\strpos(DecimalConfig::NUMERALS, $str[$i++]));

			for ( $j = 0; $j < \count($arr); $j++ )
			{
				if ( $arr[$j] > $baseOut-1 )
				{
					if ( !isset($arr[$j+1]) )
					{ $arr[$j+1] = 0; }

					$arr[$j+1] += \intval($arr[$j] / $baseOut | 0);
					$arr[$j] %= $baseOut;
				}
			}
		}

		return array_reverse($arr);
	}

	/**
	 * Convert an array of digits to a string.
	 *
	 * @param array|null $d
	 * @since 1.0.0
	 * @return string
	 */
	public static function digitsToStr (
		?array $d
	) : string
	{
		$indexOfLastWord = \is_null($d) ? 0 : \count($d)-1;
		$str = '';
		$w = $d[0]??0;

		if ( $indexOfLastWord > 0 )
		{
			$str .= \strval($w);

			for ( $i = 1; $i < $indexOfLastWord; $i++ )
			{
				$ws = $d[$i].'';
				$k = Decimal::LOG_BASE - \strlen($ws);

				if ( $k )
				{ $str .= static::getZeroString($k); }

				$str .= $ws;
			}

			$w = $d[$i];
			$ws = $w.'';
			$k = Decimal::LOG_BASE - \strlen($ws);

			if ( $k )
			{ $str .= static::getZeroString($k); }
		}
		else if ( $w === 0 )
		{ return '0'; }

		// Remove trailing zeros of last w.
		for (; $w != 0 && $w % 10 === 0;) 
		{ $w /= 10; }

		return $str.$w;
	}

	/**
	 * Calculate the base 10 exponent from 
	 * the base 1e7 exponent.
	 * It expects an array with $digits at
	 * Decimal::LOG_BASE ordened.
	 *
	 * @param array $digits
	 * @param integer $e Exponent
	 * @since 1.0.0
	 * @return integer
	 */
	public static function getBase10Exponent ( array $digits, int $e ) : int
	{
		$w = $digits[0];

		for ( $e *= Decimal::LOG_BASE; $w >= 10; $w /= 10 )
		{ $e++; }

		return $e;
	}

	/**
	 * Get precision to $digits.
	 * It expects an array with $digits at
	 * Decimal::LOG_BASE ordened.
	 * 
	 * e.g.
	 * [1250000, 2540000, ...]
	 *
	 * @param array $digits
	 * @since 1.0.0
	 * @return int
	 */
	public static function getPrecision ( array $digits ) : int
	{
		$w = \count($digits)-1;
		$len = $w * Decimal::LOG_BASE+1;

		$w = $digits[$w];

		// Non-zero
		if ( $w )
		{
			// Subtract the number of trailing zeros of the last word
			for (; $w % 10 == 0; $w /= 10) 
			{ $len--; }

			// Add the number of digits of the first word
			for ($w = $digits[0]; $w >= 10; $w /= 10)
			{ $len++; }
		}

		return $len;
	}

	/**
	 * Get an string with $k zeros.
	 *
	 * old style (too slow)
	 * 
	 * for ( ; $k--; )
	 * { $zn .= '0'; }
	 * 
	 * new style
	 * 
	 * str_repeat is faster than a for-loop
	 * 
	 * However, str_repeat becames slow with integers 
	 * greater than 999999999, then, first we decrease 
	 * 999999999 from $i till $i < 999999999, then we
	 * concat 0x999999999 $nin times and concat $i if
	 * is needed. It will give us power to concat
	 * 999999999+5000000000 of zeros.
	 * 
	 * Numbers <= 9...0.000 seconds
	 * Numbers <= 99...0.000 seconds
	 * Numbers <= 999...0.000 seconds
	 * Numbers <= 9999...0.000 seconds
	 * Numbers <= 99999...0.000 seconds
	 * Numbers <= 999999...0.001 seconds
	 * Numbers <= 9999999...0.007 seconds
	 * Numbers <= 99999999...0.037 seconds
	 * Numbers <= 999999999...0.349 seconds
	 * Numbers <= 1999999999...1.670 seconds
	 * Numbers <= 2999999999...2.432 seconds
	 * Numbers <= 3999999999...4.840 seconds
	 * Numbers <= 4999999999...5.420 seconds
	 * Numbers <= 5999999999...6.529 seconds
	 * 
	 * After 5999999999, however, memory usage is too
	 * heavy, should we fix it?
	 * 
	 * @todo 2? Try to be faster when $k is greater than 5999999999.
	 * @param integer $k Quantity of zeros.
	 * @since 1.0.0
	 * @return string
	 */
	public static function getZeroString (
		int $k
	) : string
	{ 
		$zn  = '';

		while ( $k > 999999999 )
		{
			$k = $k > PHP_INT_MAX ? \bcsub($k, 999999999) : $k-999999999;
			$zn .= str_repeat('0', 999999999);
		}
	
		if ( $k != 0 )
		{ $zn .= str_repeat('0', $k); }

		return $zn; 
	}

	/**
	 * Alias to native abs() function.
	 * But, always returning an integer.
	 *
	 * @param mixed $num
	 * @since 1.0.0
	 * @return integer
	 */
	public static function intAbs ( $num ) : int
	{ return \intval(\abs($num)); }

	/**
	 * Alias to native ceil() function.
	 * But, always returning an integer.
	 *
	 * @param float $num
	 * @since 1.0.0
	 * @return integer
	 */
	public static function intCeil ( float $num ) : int
	{ return \intval(\ceil($num)); }

	/**
	 * Alias to native floor() function.
	 * But, always returning an integer.
	 *
	 * @param float $num
	 * @since 1.0.0
	 * @return integer
	 */
	public static function intFloor ( float $num ) : int
	{ return \intval(\floor($num)); }

	/**
	 * Alias to native max() function.
	 * But, always returning an integer.
	 *
	 * @param array $nums
	 * @since 1.0.0
	 * @return integer
	 */
	public static function intMax ( array $nums ) : int
	{ return \intval(\max($nums)); }

	/**
	 * Alias to native max() function.
	 * But, always returning an integer.
	 *
	 * @param array $nums
	 * @since 1.0.0
	 * @return integer
	 */
	public static function intMin ( array $nums ) : int
	{ return \intval(\min($nums)); }

	/**
	 * Alias to native floor() function.
	 * But, always returning an integer.
	 *
	 * @param double|float|integer $base
	 * @param double|float|integer $exp
	 * @since 1.0.0
	 * @return integer
	 */
	public static function intPow ( $base, $exp ) : int
	{ return \intval(\pow($base, $exp)); }

	/**
	 * Throw an exception if $i is not a int32.
	 *
	 * @param mixed $i
	 * @param integer $min
	 * @param integer $max
	 * @since 1.0.0
	 * @return bool TRUE when is int32, throw an Exception when FALSE.
	 * @throws RuntimeException If $i is not a int32.
	 */
	public static function isInt32 ( $i, int $min, int $max ) : bool
	{
		// Handle with zeros
		if ( $i === 0 || $i === '0' || ($i === '-0' && $min <= 0) )
		{ return true; }

		if ( !\is_int($i) || $i < $min || $i > $max )
		{ throw new RuntimeException(\sprintf('`%s` is not a valid int32.', gettype($i))); }

		return true;
	}

	/**
	 * Return if number is odd or not.
	 *
	 * @param float $i
	 * @since 1.0.0
	 * @return boolean
	 */
	public static function isOdd ( $i )
	{ return $i & 1; }

	/**
	 * Enhanced native sqrt() function.
	 * 
	 * sqrt() cannot handle higher scales number,
	 * which can make some decimals calculations
	 * weird (with less precision than required).
	 * 
	 * Other side, bcsqrt() can handle with scale
	 * but cannot handle with float or exponential
	 * values.
	 * 
	 * This function analyzes $num value to determine
	 * if will be better to a bcsqrt() or, if can't, 
	 * do a sqrt() instead.
	 * 
	 * Even we are calling two Decimal functions as
	 * toString() and toNumber() it makes more fast
	 * then only do sqrt() to estimate square root.
	 * 
	 * Since some values may return INF value at
	 * sqrt() but when using bcsqrt() it will return
	 * a regular number.
	 * 
	 * But, even with this approach, sqrt() it is still
	 * the slower method to Decimals and may need some
	 * improvements soon.
	 *
	 * @param Decimal $num
	 * @param integer $scale
	 * @since 1.0.0
	 * @return Decimal
	 */
	public static function sqrt ( Decimal $num, int $scale = null ) : Decimal
	{
		$_s = Math::asStr($num);

		if ( \stripos($_s, 'e') === false && \stripos($_s, '.') === false )
		{ $s = \bcsqrt($_s, $scale+5); }
		else
		{ $s = \sqrt($num->toNumber()); }

		return new Decimal($s, $num->_c());
	}
	
	/**
	 * Exponent $e must be positive and non-zero.
	 * It does use times symbol (*) instead pow()
	 *
	 * @param integer $base
	 * @param integer $exp Integer > 0.
	 * @since 1.0.0
	 * @return int
	 */
	public static function tinyPow ( int $base, int $exp ) : int
	{
		$n = $base;

		if ( !($exp > 0) )
		{ throw new RuntimeException('Exponent $e must be positive and non-zero.'); }

		while ( --$exp )
		{ $n *= $base; }
		
		return $n;
	}

	/**
	 * Compare two arrays $a and $b. By checking
	 * if $a has any value greater than $b.
	 *
	 * @param array $a
	 * @param array $b
	 * @param integer $al $a length
	 * @param integer $bl $b length
	 * @since 1.0.0
	 * @return integer
	 */
	public static function compare ( 
		array $a, 
		array $b, 
		int $al, 
		int $bl 
	) : int
	{
		if ( $al != $bl )
		{ $r = $al > $bl ? 1 : -1; }
		else
		{
			for ( $i = $r = 0; $i < $al; $i++ )
			{
				if ( $a[$i] != $b[$i] )
				{
					$r = $a[$i] > $b[$i] ? 1 : -1;
					break;
				}
			}
		}

		return $r;
	}

	/**
	 * Used by divide() function.
	 * Multiplies all integers at $x array by $k adding $carry
	 * and, after, normalize it to $base.
	 * 
	 * Assumes non-zero x and k, and hence non-zero result.
	 *
	 * @param array $x
	 * @param integer $k
	 * @param integer $base
	 * @since 1.0.0
	 * @return array
	 */
	public static function multiplyInt ( 
		array $x, 
		int $k, 
		int $base 
	) : array
	{
		$carry = 0;
		$i = \count($x);

		for ( $x = $x; $i--; )
		{
			$temp = $x[$i] * $k + $carry;
			$x[$i] = $temp % $base | 0;
			$carry = $temp / $base | 0;
		}

		if ( $carry )
		{ \array_unshift($x, $carry); }

		return $x;
	}

	/**
	 * Used by divide() function.
	 * Subtract $b from $a.
	 *
	 * @param array $a
	 * @param array $b
	 * @param integer $al $a length
	 * @param integer $base
	 * @since 1.0.0
	 * @return array
	 */
	public static function subtract ( 
		array $a, 
		array $b, 
		int $al, 
		int $base 
	) : array
	{
		$i = 0;

		for (; $al--;)
		{
			$a[$al] -= $i;
			$i = isset($a[$al], $b[$al]) && $a[$al] < $b[$al] ? 1 : 0;
			$a[$al] = $i * $base + ($a[$al]??0) - ($b[$al]??0);
		}

		for(; !$a[0] && \count($a)-1; )
		{ \array_shift($a); }

		return $a;
	}
}